pkg/cloudprovider/vsphereparavirtual/route.go

```diff
@@ -0,0 +1,415 @@
+	AnnotationClusterName = v1alpha1.GroupName + "/clusterName"
+	// RealizedStateTimeout is the timeout duration for realized state check
+	// RealizedStateTimeout = 10 * time.Second
+	// RealizedStateSleepTime is the interval between realized state check
+	// RealizedStateSleepTime = 1 * time.Second
+)
+
+var updateNetworkConditionBackoff = wait.Backoff{
+	Steps:    5,
+	Duration: 100 * time.Millisecond,
+	Jitter:   1.0,
+}
+
+// A list of possible RouteSet operation error messages
+var (
+	ErrGetRouteSet    = errors.New("failed to get RouteSet")
+	ErrCreateRouteSet = errors.New("failed to create RouteSet")
+	ErrListRouteSet   = errors.New("failed to list RouteSet")
+	ErrDeleteRouteSet = errors.New("failed to delete RouteSet")
+)
+
+// GetRouteSetClient returns a new RouteSet client that can be used to access SC
+func GetRouteSetClient(config *rest.Config) (client.Interface, error) {
+	v1alpha1.AddToScheme(scheme.Scheme)
+	rClient, err := client.NewForConfig(config)
+	if err != nil {
+		klog.V(6).Infof("Failed to create RouteSet clientset")
+		return nil, err
+	}
+	return rClient, nil
+}
+
+// NewRoutes returns an implementation of RoutesProvider
+func NewRoutes(clusterNS string, kcfg *rest.Config, ownerRef metav1.OwnerReference, kubeClient clientset.Interface, nodeLister listerv1.NodeLister) (RoutesProvider, error) {
+	routeClient, err := GetRouteSetClient(kcfg)
+	if err != nil {
+		return nil, err
+	}
+
+	ownerRefs := []metav1.OwnerReference{
+		ownerRef,
+	}
+
+	return &routesProvider{
+		routeClient: routeClient,
+		namespace:   clusterNS,
+		nodeMap:     make(map[string]*v1.Node),
+		ownerRefs:   ownerRefs,
+		kubeClient:  kubeClient,
+		nodeLister:  nodeLister,
+	}, nil
+}
+
+// ListRoutes implements Routes.ListRoutes
+// Get RouteSet CR from SC namespace and then filters routes that belong to the specified clusterName
+// Only return cloudprovider.Route if RouteSet CR status 'Ready' is true
+func (r *routesProvider) ListRoutes(ctx context.Context, clusterName string) ([]*cloudprovider.Route, error) {
+	klog.V(6).Infof("Attempting to list Routes for cluster %s", clusterName)
+
+	routeSets, err := r.routeClient.NsxV1alpha1().RouteSets(r.namespace).List(ctx, metav1.ListOptions{})
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return []*cloudprovider.Route{}, nil
+		}
+		klog.ErrorS(ErrListRouteSet, fmt.Sprintf("%v", err))
+		return nil, err
+	}
+	if len(routeSets.Items) == 0 {
+		return []*cloudprovider.Route{}, nil
+	}
+	return r.createCPRoutes(routeSets, clusterName), nil
+}
+
+// createCPRoutes creates cloudprovider Routes based on RouteSet CR
+func (r *routesProvider) createCPRoutes(routeSets *v1alpha1.RouteSetList, clusterName string) []*cloudprovider.Route {
+	var routes []*cloudprovider.Route
+	for _, routeSet := range routeSets.Items {
+		routeSetClusterName := routeSet.Annotations[AnnotationClusterName]
+		if routeSetClusterName == clusterName {
+			// only return cloudprovider.Route if RouteSet CR status 'Ready' is true
```

Comment:
I commented the check for the realized state of routes for now, we could either 1. uncomment & merge them now, and fix/test later once backend finishes, or 2. uncomment only after we test the logic after backend finishes

Reply:
I don't think you need to wait for backend, you can always check the CR status.
Or do you mean after we add realized routes in CR status, you'll switch to get realized routes there?

Reply:
ah I mean after backend finishes, I could test the logic and verify if it's working correctly, but I could also merge this logic before testing it with backend, if there are any bugs I could fix later.

Reply:
nvm, I uncomment the code, we could merge them together